File: go/recursive/valid_bst.go
package main

import (
	"math"
	"reflect"
	"sort"
)

//验证二叉搜索树

//根据二叉搜索树的特性，进行递归判断
func isValidBST(root *TreeNode) bool {
	return validBstHelper(root, math.MinInt64, math.MaxInt64)
}

func validBstHelper(node *TreeNode, lower, upper int) bool {
	if node == nil {
		return true
	}

	if node.Val <= lower || node.Val >= upper {
		return false
	}
	return validBstHelper(node.Left, lower, node.Val) && validBstHelper(node.Right, node.Val, upper)
}

//中序遍历二叉树，然后看是否有序
func isValidBST2(root *TreeNode) bool {
	if root == nil {
		return true
	}

	set := map[int]bool{}
	var temp []int
	iterative(root, set, &temp)

	var data []int
	for v := range set {
		data = append(data, v)
	}
	sort.Ints(data)
	return reflect.DeepEqual(temp, data)
}

func iterative(node *TreeNode, data map[int]bool, temp *[]int) {
	if node != nil {
		iterative(node.Left, data, temp)
		data[node.Val] = true
		*temp = append(*temp, node.Val)
		iterative(node.Right, data, temp)
	}
}

//中序遍历优化，不用全部将结果保存，只需要每次比较前后节点的值即可

func isValidBST3(root *TreeNode) bool {
	var stack []*TreeNode
	var prev *TreeNode
	cur := root
	for cur != nil || len(stack) > 0 {
		for cur != nil {
			stack = append(stack, cur)
			cur = cur.Left
		}
		node := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		if prev != nil && prev.Val >= node.Val {
			return false
		}
		prev = node
		cur = node.Right
	}
	return true
}

File: cpp/tree/valid_bst.cpp
// 验证二叉搜索树
#include "vector"
#include "stdio.h"
#include "algorithm"
#include "set"

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {

private:
    TreeNode *prev;

public:

    // 递归的方式
    bool isValidBST(TreeNode* root) {

    }

    // 中序遍历的方式，先遍历一遍，将数据存储在数组中，再排序进行比较
    bool isValidBST1(TreeNode* root) {
        std::set<int> set;
        std::vector<int> data;
        helper(root, set, data);

        std::vector<int> temp;
        for (int i : set) {
            temp.push_back(i);
        }
        std::sort(temp.begin(), temp.end());
        return temp == data;
    }

    void helper(TreeNode *root, std::set<int>& set, std::vector<int>& data) {
        if (root == nullptr) {
            return;
        }
        helper(root->left, set, data);
        set.insert(root->val);
        data.push_back(root->val);
        helper(root->right, set, data);
    }

    // 中序遍历优化，比较前后节点
    bool isValidBST2(TreeNode* root) {
        if (root == nullptr) {
            return true;
        }

        if (!isValidBST2(root->left)) {
            return false;
        }
        if (prev != nullptr && prev->val >= root->val) {
            return false;
        }
        prev = root;
        return isValidBST2(root->right);
    }
};
